lms/djangoapps/grades/models.py

```diff
@@ -0,0 +1,42 @@
+from django.contrib.auth.models import User
+from django.db import models
+
+from xmodule_django.models import BlockTypeKeyField, CourseKeyField, UsageKeyField
+
+class Grade(models.Model):
+    """
+    A django model tracking persistent grades at the subsection level.
+    """
+
+    class Meta(object):
+        index_together = [
+            ['user_id', 'course_id', 'content_type', 'is_valid'],
+            ['course_id', 'updated'],
+            ['updated'],
+        ]
+
+    # TODO: timestamp discussions - do we need all three of [created, updated, subtree_edited_date]?
+    created = models.DateTimeField('creation timestamp', auto_now_add=True)
+    updated = models.DateTimeField('latest updated timestamp', auto_now=True)
+    subtree_edited_date = models.DateTimeField('last content edit timestamp')
+
+    # TODO: is this the right way to store course edit version? make sure it matches up w/ what Nimisha's exposing in her PR
+    course_version = models.CharField('guid of latest course version', max_length=255)
+
+    user = models.ForeignKey(User)
+    course_id = CourseKeyField(max_length=255)
+    content_id = UsageKeyField(max_length=255)
+    content_type = models.CharField('used to bucket grades by item type', max_length=255)
+    total_weighted_raw_score = models.IntegerField()
+    total_weighted_max_score = models.IntegerField()
+    is_valid = models.BinaryField()
+    visible_blocks = models.ManyToManyField(BlockRecord)  # see TODO below
```

Comment:
Ah, I misunderstood the proposed implementation. A M:M field here will really blow things out, since that'll make a join table multiplying potentially 100M records X 10-ish items per blocks. I thought what was being proposed was a simple foreign key from `Grade` to some new table, where that table would have in it:

1. The `content_id`/`block_id` of the top level thing being graded (so a sequence)
2. A text field with a list of all the visible blocks underneath.
3. Some hash of #2 so that we can do indexed lookups and add permutations as necessary.
4. Other metadata if it's useful.